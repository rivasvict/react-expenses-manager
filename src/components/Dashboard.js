import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Results from './Results';
import AddEntry from './AddEntry';
import Summary from './Summary';
import EntrySummaryWithFilter from './EntrySummaryWithFilter'
import { getEntryCategoryOption } from '../helpers/entriesHelper'; 

import { Switch, Route, Link } from 'react-router-dom';

const dasahboardStyle = {
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100vh'
};

function Dashboard({ entries, match }) {
  return (
    <div style={dasahboardStyle}>
      <Switch>
        <Route exact path={`${match.url}`}>
          <Link to={`${match.url}/add-income`}>Add income</Link>
          <Results 
            entries={entries}
            baseUrl={match.url}/>
          <Link to={`${match.url}/add-expense`}>Add expense</Link>
        </Route>
        <Route path={`${match.url}/add-income`}>
          <AddEntry entryType='income' />
        </Route>
        <Route path={`${match.url}/add-expense`}>
          <AddEntry entryType='outcome' />
        </Route>
        <Route path={`${match.url}/incomes`}>
          <EntrySummaryWithFilter categoryOptions={getEntryCategoryOption('income')} entries={entries['incomes']} name='Incomes' />
        </Route>
        <Route path={`${match.url}/outcomes`}>
          <EntrySummaryWithFilter categoryOptions={getEntryCategoryOption('outcome')} entries={entries['outcomes']} name='Outcomes'/>
        </Route>
        <Route path={`${match.url}/summary`}>
          <Summary entries={entries} />
        </Route>
      </Switch>
    </div>
  )
}

const mapStateToProps = state => ({
  entries: state.expensesManager.entries
});

Dashboard.propTypes = {
  entries: PropTypes.shape({
    incomes: PropTypes.arrayOf(
      PropTypes.shape({
        ammount: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        category: PropTypes.string.isRequired
      }).isRequired
    ).isRequired,
    outcomes: PropTypes.arrayOf(
      PropTypes.shape({
        ammount: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        category: PropTypes.string.isRequired
      }).isRequired
    ).isRequired
  }).isRequired
};

export default connect(mapStateToProps)(Dashboard);
