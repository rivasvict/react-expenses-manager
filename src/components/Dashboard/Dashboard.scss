@import '../../variables.scss';

.main-container {
  height: 100vh;
  padding-bottom: 1.875rem;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
}