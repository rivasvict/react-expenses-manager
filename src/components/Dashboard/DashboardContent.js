import React from 'react';
import { Button, Col, Container, Row } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import Results from '../Results';
import { getMonthNameDisplay } from '../../helpers/date';
import { getNewSelectedDate, doesAdjacentDateExist, calculateTotal } from '../../helpers/general';
import ScreenTitle from '../common/ScreenTitle';
import { connect } from 'react-redux';
import { setSelectedDate } from '../../redux/expensesManager/actionCreators';
import { formatNumberForDisplay, getSum } from '../../helpers/entriesHelper/entriesHelper';
import './DashboardContent.scss';
import { IconRemote } from '../common/Icons';
import ContentTileSection from '../common/ContentTitleSection';

const handleDateSelectionPointers = ({ entries, selectedDate, onSelectedDateChange, dateAdjacencyType }) => (
  onSelectedDateChange(getNewSelectedDate({
    entries,
    dateAdjacencyType,
    currentSelectedDate: selectedDate
  }))
);

const DashboardContent = ({ entries, match, selectedDate, onSelectedDateChange }) => {
  const monthBalance = (entries[selectedDate.year] && entries[selectedDate.year][selectedDate.month]) || { incomes: [], expenses: [] };
  const summaryUrl = `${match.url}/summary`;
  const incomesName = 'incomes';
  const expensesName = 'expenses';
  const incomesSum = getSum({ entryType: incomesName, entries: monthBalance })
  const expensesSum = getSum({ entryType: expensesName, entries: monthBalance })
  const totalSum = calculateTotal(incomesSum, expensesSum);

  return (
    <Container className='DasboardContent'>
      <ContentTileSection title='Summary' to={summaryUrl}>
        {`Savings `}<IconRemote inLine={true} />{` ${formatNumberForDisplay(totalSum)}`}
      </ContentTileSection>
      <Row className='month-header'>
        <Col xs={3}>
          {
            doesAdjacentDateExist({ dateAdjacencyType: 'prev', selectedDate, entries }) ?
            <Button onClick={() => handleDateSelectionPointers({ entries, selectedDate, onSelectedDateChange, dateAdjacencyType: 'prev' })}>Prev</Button> :
            null
          }
        </Col>
        <Col xs={6}>
          <ScreenTitle screenTitle={`${getMonthNameDisplay(selectedDate.month)} ${selectedDate.year}`} />
        </Col>
        <Col xs={3}>
          {
            doesAdjacentDateExist({ dateAdjacencyType: 'next', selectedDate, entries }) ?
            <Button onClick={() => handleDateSelectionPointers({ entries, selectedDate, onSelectedDateChange, dateAdjacencyType: 'next' })}>Next</Button> :
            null
          }
        </Col>
      </Row>
      <MonthContent {...{ entries: monthBalance, match }} />
    </Container>
  );
};

const MonthContent = ({ entries, match }) => (
  <React.Fragment>
    <Row className='top-container'>
      <Col xs={12} className='top-content'>
        <Results
          entries={entries}
          baseUrl={match.url} />
      </Col>
    </Row>
    <Row>
      <Col xs={12} className='bottom-content'>
        <Link to={`${match.url}/add-income`} className='btn btn-primary btn-block'>Add Income</Link>
        <Link to={`${match.url}/add-expense`} className='btn btn-secondary btn-block'>Add Expenses</Link>
      </Col>
    </Row>
  </React.Fragment>
);

const mapStateToProps = state => ({
  entries: state.expensesManager.entries,
  selectedDate: state.expensesManager.selectedDate
});

const mapActionsToProps = dispatch => ({
  onSelectedDateChange: newSelectedDate => dispatch(setSelectedDate(newSelectedDate))
});

export default connect(mapStateToProps, mapActionsToProps)(DashboardContent)