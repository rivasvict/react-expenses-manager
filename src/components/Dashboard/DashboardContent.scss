@import '../../variables.scss';

.dashboard-content-container {
  background-color: $color_clear_4;
  border-radius: $standard-radius;
}