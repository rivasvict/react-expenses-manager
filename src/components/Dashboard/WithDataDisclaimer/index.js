import React, { useState } from "react";
import DataDisclaimerModal from "./components/DataDisclaimer";
import everShowDataDisclaimerStorage from "./utils/everShowDataDisclaimerStorage";

const WithDataDisclaimer = ({ children }) => {
  const everShowDataDisclaimer = everShowDataDisclaimerStorage.get();

  const [showDataDisclaimer, setShowDataDisclaimer] = useState(
    everShowDataDisclaimer
  );

  const hideDataDisclaimer = ({ everShowDataDisclaimer }) => {
    everShowDataDisclaimerStorage.set(everShowDataDisclaimer);
    setShowDataDisclaimer(false);
  };

  return (
    <>
      <DataDisclaimerModal
        show={showDataDisclaimer}
        onHide={hideDataDisclaimer}
      />
      {children}
    </>
  );
};

export default WithDataDisclaimer;
