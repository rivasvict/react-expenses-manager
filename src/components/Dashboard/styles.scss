@import '../../variables.scss';

.main-container {
  height: 100vh;
  /**
  * Look at src/helpers/general.js's function `addViewHeightMobileConfig`
  * To understand how we calculate this `--vh` var
  */
  height: calc(var(--vh, 1vh) * 100);
  padding-bottom: 1.875rem;
  display: flex;
  flex-direction: column;
  min-height: fit-content;
  background-color: #292c32;

  .dashboard-container {
    flex-grow: 1;
  }
}
