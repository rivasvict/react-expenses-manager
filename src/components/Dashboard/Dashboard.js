import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Results from '../Results';
import AddEntry from '../common/ExpensesManager/AddEntry/AddEntry';
import Summary from '../../components/common/ExpensesManager/Summaries/Summary';
import EntrySummaryWithFilter from '../common/ExpensesManager/Summaries/EntrySummaryWithFilter'

import { Switch, Route, Link, useRouteMatch } from 'react-router-dom';
import { Col, Container, Row } from 'react-bootstrap';
import Header from '../common/Header';
import './Dashboard.scss';
import WorkAreaContentContainer from '../common/WorkAreaContentContainer';
import { getBalance } from '../../redux/expensesManager/actionCreators';

const DashboardContent = ({ entries, match }) => (
  <React.Fragment>
    <Row className='top-container'>
      <Col xs={12} className='top-content'>
        <Results
          entries={entries}
          baseUrl={match.url} />
      </Col>
    </Row>
    <Row>
      <Col xs={12} className='bottom-content'>
        <Link to={`${match.url}/add-income`} className='btn btn-primary btn-block'>Add Income</Link>
        <Link to={`${match.url}/add-expense`} className='btn btn-secondary btn-block'>Add Expenses</Link>
      </Col>
    </Row>
  </React.Fragment>
);

function Dashboard({ entries }) {
  const match = useRouteMatch(); 
 
  return (
    <main className='main-container'>
      <Header />
      <Container fluid>
        <WorkAreaContentContainer>
          <Switch>
            <Route path={`${match.url}/add-income`}>
              <AddEntry entryType='income' />
            </Route>
            <Route path={`${match.url}/add-expense`}>
              <AddEntry entryType='expense' />
            </Route>
            <Route path={`${match.url}/incomes`}>
              <EntrySummaryWithFilter entryType='income' />
            </Route>
            <Route path={`${match.url}/expenses`}>
              <EntrySummaryWithFilter entryType='expense' />
            </Route>
            <Route path={`${match.url}/summary`}>
              <Summary entries={entries} />
            </Route>
            <Route exact path={`${match.url}`}>
              <DashboardContent {...{ entries, match }} />
            </Route>
          </Switch>
        </WorkAreaContentContainer>
      </Container>
    </main>
  )
}

const mapStateToProps = state => ({
  entries: state.expensesManager.entries
});

Dashboard.propTypes = {
  entries: PropTypes.shape({
    incomes: PropTypes.arrayOf(
      PropTypes.shape({
        amount: PropTypes.number.isRequired,
        description: PropTypes.string.isRequired,
        category: PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired,
          type: PropTypes.string.isRequired
        })
      }).isRequired
    ).isRequired,
    expenses: PropTypes.arrayOf(
      PropTypes.shape({
        amount: PropTypes.number.isRequired,
        description: PropTypes.string.isRequired,
        category: PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired,
          type: PropTypes.string.isRequired
        })
      }).isRequired
    ).isRequired
  }).isRequired
};

export default connect(mapStateToProps)(Dashboard);
