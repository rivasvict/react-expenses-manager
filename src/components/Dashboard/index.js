import React, { useEffect, useRef } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import AddEntry from "../common/ExpensesManager/AddEntry";
import Summary from "../common/ExpensesManager/Summaries/Summary";
import EntrySummaryWithFilter from "../common/ExpensesManager/Summaries/EntriesSummaryWithFilter";

import { Switch, Route, useRouteMatch } from "react-router-dom";
import { Container } from "react-bootstrap";
import Header from "../common/Header";
import "./styles.scss";
import WorkAreaContentContainer from "../common/WorkAreaContentContainer";
import DashboardContent from "./components/DashboardContent";
import { addViewHeightMobileConfig } from "../../helpers/general";
import EditEntry from "../common/ExpensesManager/EditEntry";

function Dashboard({ entries, selectedDate }) {
  useEffect(() => {
    /**
     * TODO: Make sure this calculation also occurs when resizing the screen
     * as well and implement a debounced call.
     * https://github.com/rivasvict/react-expenses-manager/issues/63
     */
    addViewHeightMobileConfig();
  }, []);
  const mainRef = useRef(null);
  const match = useRouteMatch();

  return (
    <main className="main-container container-background" ref={mainRef}>
      <Header />
      <Container fluid className="dashboard-container">
        <WorkAreaContentContainer>
          <Switch>
            <Route path={`${match.url}add-income`}>
              <AddEntry entryType="income" selectedDate={selectedDate} />
            </Route>
            <Route path={`${match.url}edit-income/:entryId`}>
              <EditEntry entryType="income" selectedDate={selectedDate} />
            </Route>
            <Route path={`${match.url}add-expense`}>
              <AddEntry entryType="expense" selectedDate={selectedDate} />
            </Route>
            <Route path={`${match.url}edit-expense/:entryId`}>
              <EditEntry entryType="income" selectedDate={selectedDate} />
            </Route>
            <Route path={`${match.url}incomes`}>
              <EntrySummaryWithFilter
                selectedDate={selectedDate}
                entryType="income"
              />
            </Route>
            <Route path={`${match.url}expenses`}>
              <EntrySummaryWithFilter
                selectedDate={selectedDate}
                entryType="expense"
              />
            </Route>
            <Route path={`${match.url}summary`}>
              {/* TODO: Fix the issue that appears when the screen is refreshed on the summary route */}
              <Summary entries={entries} selectedDate={selectedDate} />
            </Route>
            <Route path={`${match.url}dashboard`}>
              <DashboardContent {...{ entries, match }} />
            </Route>
            <Route exact path={`${match.url}`}>
              <DashboardContent {...{ entries, match }} />
            </Route>
          </Switch>
        </WorkAreaContentContainer>
      </Container>
    </main>
  );
}

const mapStateToProps = (state) => ({
  entries: state.expensesManager.entries,
  selectedDate: state.expensesManager.selectedDate,
});

// TODO: Fix this propTypes and add it through the whole application
Dashboard.propTypes = {
  entries: PropTypes.shape({
    incomes: PropTypes.arrayOf(
      PropTypes.shape({
        amount: PropTypes.number.isRequired,
        description: PropTypes.string,
        timestamp: PropTypes.number.isRequired,
        type: PropTypes.string.isRequired,
        category: PropTypes.shape({
          id: PropTypes.number,
          name: PropTypes.string.isRequired,
        }),
      }).isRequired
    ).isRequired,
    expenses: PropTypes.arrayOf(
      PropTypes.shape({
        amount: PropTypes.number.isRequired,
        description: PropTypes.string,
        timestamp: PropTypes.number.isRequired,
        type: PropTypes.string.isRequired,
        category: PropTypes.shape({
          id: PropTypes.number,
          name: PropTypes.string.isRequired,
        }),
      }).isRequired
    ).isRequired,
  }).isRequired,
};

export default connect(mapStateToProps)(Dashboard);
