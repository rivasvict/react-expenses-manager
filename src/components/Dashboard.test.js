import React from 'react';
import Dashboard from './Dashboard';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route } from 'react-router-dom';
// TODO: Test routes

describe('Dashboard component test', () => {
  let jsDomTree = null;
  let DashboardInstance = null;
  let newEntry = null;

  beforeEach(() => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={['/']}><Route component={Dashboard} path='/'></Route></MemoryRouter>);
    jsDomTree = tree.toJSON();
    DashboardInstance = tree.root.findByType(Dashboard).instance;
    newEntry = {
      ammount: '12',
      description: 'fwwr',
      type: 'income',
      category: 'salary'
    };
  });

  it('Renders Dashboard component properly', () => {
    expect(jsDomTree).toMatchSnapshot();
  });

  it('addEntry: Add new entry to state', () => {
    act(() => {
      DashboardInstance.addEntry(newEntry, 'incomes');
      DashboardInstance.addEntry(newEntry, 'incomes');
    });

    expect(DashboardInstance.state.entries.incomes.length).toBe(2);
    expect(DashboardInstance.state.entries.incomes).toStrictEqual([ newEntry, newEntry ]);
  })

  it('addIncome: should add new income to the state outcomes', () => {
    act(() => {
      DashboardInstance.addIncome(newEntry);
      DashboardInstance.addIncome(newEntry);
    });

    expect(DashboardInstance.state.entries.incomes.length).toBe(2);
    expect(DashboardInstance.state.entries.incomes).toStrictEqual([ newEntry, newEntry ]);
  })
  
  it('addOutcome: should add new outcome to the state outcomes', () => {
    act(() => {
      DashboardInstance.addOutcome(newEntry);
      DashboardInstance.addOutcome(newEntry);
    });

    expect(DashboardInstance.state.entries.outcomes.length).toBe(2);
    expect(DashboardInstance.state.entries.outcomes).toStrictEqual([ newEntry, newEntry ]);
  })

  it('getSum: should get the correct sum for entries added', () => {
    act(() => {
      DashboardInstance.addOutcome(newEntry);
      DashboardInstance.addOutcome(newEntry);
      DashboardInstance.addOutcome(newEntry);
      DashboardInstance.addOutcome(newEntry);
    });

    const totalSum = DashboardInstance.getSum('outcomes');
    expect(totalSum).toBe(parseInt(newEntry.ammount) * 4);
  })
})
