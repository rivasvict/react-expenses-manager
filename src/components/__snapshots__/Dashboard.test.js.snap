// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component test Renders Dashboard component properly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "100vh",
      "justifyContent": "center",
    }
  }
>
  <a
    href="/add-income"
    onClick={[Function]}
  >
    Add income
  </a>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <a
      href="/incomes"
      onClick={[Function]}
    >
      <div
        style={
          Object {
            "border": "solid 1px black",
            "padding": "20px 20px",
          }
        }
      >
        Income
        <br />
        0
      </div>
    </a>
    <a
      href="/outcomes"
      onClick={[Function]}
    >
      <div
        style={
          Object {
            "border": "solid 1px black",
            "padding": "20px 20px",
          }
        }
      >
        Exoepenses
        <br />
        0
      </div>
    </a>
    <a
      href="/summary"
      onClick={[Function]}
    >
      <div
        style={
          Object {
            "border": "solid 1px black",
            "padding": "20px 20px",
          }
        }
      >
        Total
        <br />
        0
      </div>
    </a>
  </div>
  <a
    href="/add-expense"
    onClick={[Function]}
  >
    Add expense
  </a>
</div>
`;
